File: backend/core/ranking.py
import math
import re
from config import Config

config = Config()

def normalize_pagerank(pagerank_score):
    if pagerank_score is None or pagerank_score <= 0:
        return 0.0
    # CHANGED: Removed the ** 0.8 smoothing. 
    # We want a linear or stricter curve to punish low-authority pages more heavily.
    normalized = pagerank_score / 100.0
    return normalized

def normalize_pageviews(pageview_count):
    if pageview_count is None or pageview_count < 1:
        return 0.0
    
    # Log scale is good, but let's shift the floor.
    # Articles with < 100 views/month are usually noise.
    if pageview_count < 100:
        return 0.1

    min_log = 2.0 # 100 views
    max_log = 7.0 # 10M views
    
    log_val = math.log10(pageview_count)
    score = (log_val - min_log) / (max_log - min_log)
    return min(1.0, max(0.0, score))

def calculate_title_match_score(title: str, query: str) -> float:
    title_lower = title.lower().replace('_', ' ')
    query_lower = query.lower()
    
    title_words = set(title_lower.split())
    query_words = set(query_lower.split())
    
    if not title_words:
        return 0.0
    
    intersection = len(title_words & query_words)
    union = len(title_words | query_words)
    
    if union == 0:
        return 0.0
    
    base_score = intersection / union
    
    if title_lower == query_lower:
        return 1.0
    elif title_lower.startswith(query_lower) or query_lower in title_lower:
        base_score = min(1.0, base_score * 1.5)
    
    if " in " in title_lower:
        parts = title_lower.split(" in ")
        if len(parts) == 2:
            place_indicators = [
                "africa", "asia", "europe", "america", "states", "kingdom",
                "china", "india", "russia", "france", "germany", "japan",
                "canada", "australia", "brazil", "mexico", "italy", "spain",
                "california", "texas", "york", "london", "paris", "tokyo"
            ]
            if any(place in parts[1] for place in place_indicators):
                base_score *= 0.5
    
    if re.match(r'^\d{4}', title_lower):
        base_score *= 0.4
    
    meta_prefixes = ["list of", "index of", "glossary of", "timeline of", 
                     "outline of", "history of", "bibliography of"]
    if any(title_lower.startswith(prefix) for prefix in meta_prefixes):
        base_score *= 0.1 # Heavily penalize lists
    
    return min(1.0, max(0.0, base_score))

def is_meta_page(title):
    lower = title.lower()
    bad_prefixes = [
        'wikipedia:', 'template:', 'category:', 'portal:', 'help:', 
        'user:', 'talk:', 'file:', 'mediawiki:', 'draft:'
    ]
    return any(lower.startswith(p) for p in bad_prefixes) or '(disambiguation)' in lower

def calculate_multisignal_score(semantic_similarity, pagerank_score, pageview_count, title, query):
    sem_norm = float(semantic_similarity)
    pr_norm = normalize_pagerank(pagerank_score)
    pv_norm = normalize_pageviews(pageview_count)
    title_norm = calculate_title_match_score(title, query)
    
    sem_norm = max(sem_norm, config.EPSILON)
    pr_norm = max(pr_norm, config.EPSILON)
    pv_norm = max(pv_norm, config.EPSILON)
    title_norm = max(title_norm, config.EPSILON)
    
    # Geometric Mean
    score = (
        (sem_norm ** config.WEIGHT_SEMANTIC) *
        (pr_norm ** config.WEIGHT_PAGERANK) *
        (pv_norm ** config.WEIGHT_PAGEVIEWS) *
        (title_norm ** config.WEIGHT_TITLE_MATCH)
    )
    
    # OBSCURITY PENALTY:
    # If an article is semantically relevant but has near-zero popularity,
    # crush its score to prevent "random bullshit" from appearing.
    if pv_norm < 0.2 and pr_norm < 0.1:
        score *= 0.5

    return score

File: rs/src/search/ranking.rs
use crate::config::get_config;
use regex::Regex;
use std::collections::HashSet;
use std::sync::OnceLock;

// Pre-compiled regex for performance
static YEAR_REGEX: OnceLock<Regex> = OnceLock::new();

fn get_year_regex() -> &'static Regex {
    YEAR_REGEX.get_or_init(|| Regex::new(r"^\d{4}").unwrap())
}

pub fn normalize_pagerank(pagerank_score: Option<f64>) -> f64 {
    match pagerank_score {
        Some(score) if score > 0.0 => score / 100.0,
        _ => 0.0,
    }
}

pub fn normalize_pageviews(pageview_count: Option<i64>) -> f64 {
    let count = match pageview_count {
        Some(c) if c > 0 => c as f64,
        _ => return 0.0,
    };

    if count < 100.0 {
        return 0.1;
    }

    let min_log = 2.0; // 100 views
    let max_log = 7.0; // 10M views

    let log_val = count.log10();
    let score = (log_val - min_log) / (max_log - min_log);
    
    // Clamp between 0.0 and 1.0
    score.max(0.0).min(1.0)
}

pub fn calculate_title_match_score(title: &str, query: &str) -> f64 {
    let title_lower = title.to_lowercase().replace('_', " ");
    let query_lower = query.to_lowercase();

    let title_words: HashSet<&str> = title_lower.split_whitespace().collect();
    let query_words: HashSet<&str> = query_lower.split_whitespace().collect();

    if title_words.is_empty() {
        return 0.0;
    }

    let intersection_count = title_words.intersection(&query_words).count();
    let union_count = title_words.union(&query_words).count();

    if union_count == 0 {
        return 0.0;
    }

    let mut base_score = intersection_count as f64 / union_count as f64;

    // Exact or substring match boost
    if title_lower == query_lower {
        return 1.0;
    } else if title_lower.starts_with(&query_lower) || title_lower.contains(&query_lower) {
        base_score = (base_score * 1.5).min(1.0);
    }

    // Place indicator penalty logic
    // Checks if " in " splits the title and the second part contains a location
    if let Some((_, suffix)) = title_lower.split_once(" in ") {
        let place_indicators = [
            "africa", "asia", "europe", "america", "states", "kingdom",
            "china", "india", "russia", "france", "germany", "japan",
            "canada", "australia", "brazil", "mexico", "italy", "spain",
            "california", "texas", "york", "london", "paris", "tokyo"
        ];
        if place_indicators.iter().any(|&place| suffix.contains(place)) {
            base_score *= 0.5;
        }
    }

    // Year penalty
    if get_year_regex().is_match(&title_lower) {
        base_score *= 0.4;
    }

    // Meta page penalty
    let meta_prefixes = [
        "list of", "index of", "glossary of", "timeline of", 
        "outline of", "history of", "bibliography of"
    ];
    if meta_prefixes.iter().any(|&prefix| title_lower.starts_with(prefix)) {
        base_score *= 0.1;
    }

    base_score.max(0.0).min(1.0)
}

pub fn is_meta_page(title: &str) -> bool {
    let lower = title.to_lowercase();
    let bad_prefixes = [
        "wikipedia:", "template:", "category:", "portal:", "help:", 
        "user:", "talk:", "file:", "mediawiki:", "draft:"
    ];
    
    bad_prefixes.iter().any(|&p| lower.starts_with(p)) || lower.contains("(disambiguation)")
}

pub fn calculate_multisignal_score(
    semantic_similarity: f32,
    pagerank_score: f64,
    pageview_count: f64,
    title: &str,
    query: &str,
) -> f64 {
    let config = get_config();

    let sem_norm = (semantic_similarity as f64).max(config.epsilon);
    let pr_norm = pagerank_score.max(config.epsilon);
    let pv_norm = pageview_count.max(config.epsilon);
    let title_norm = calculate_title_match_score(title, query).max(config.epsilon);

    // Geometric Mean
    let mut score = sem_norm.powf(config.weight_semantic) *
                    pr_norm.powf(config.weight_pagerank) *
                    pv_norm.powf(config.weight_pageviews) *
                    title_norm.powf(config.weight_title_match);

    // Obscurity Penalty
    // If semantically relevant but near-zero popularity, crush score
    if pv_norm < 0.2 && pr_norm < 0.1 {
        score *= 0.5;
    }

    score
}